File: backend/internal/games/tanks/replay.go
package tanks

import (
	"bytes"
	"compress/gzip"
	"encoding/json"
	"io"
	"math"
)

// ReplayPlayer describes one slot's participant for a replay's header.
type ReplayPlayer struct {
	Slot    int    `json:"slot"`
	Name    string `json:"name"`
	BotID   string `json:"bot_id,omitempty"`
	Version int    `json:"version,omitempty"`
	House   bool   `json:"house"`
	Source  string `json:"source,omitempty"`
}

// Frame is one tick's recorded state, coordinates rounded to 0.01 and
// angles to 0.001 to keep replays small.
type Frame struct {
	T int         `json:"t"`
	K [][]float64 `json:"k"` // per slot: x, y, hull, turret, hp, reload, alive(0|1)
	S [][]float64 `json:"s"` // id, owner, x, y
	B [][]float64 `json:"b"` // active bonuses: x, y
	Z float64     `json:"z"` // zone radius
}

// PlayerResult is one slot's final outcome.
type PlayerResult struct {
	Slot      int    `json:"slot"`
	Place     int    `json:"place"`
	Kills     int    `json:"kills"`
	Damage    int    `json:"damage"`
	DeathTick *int   `json:"death_tick"`
	Status    string `json:"status"` // ok | crashed | timeout | invalid
}

// Replay is the full recorded match: everything needed to play it back
// without re-simulating it.
type Replay struct {
	Version  int            `json:"version"`
	Engine   string         `json:"engine"`
	Seed     int64          `json:"seed"`
	Map      string         `json:"map"`
	TickRate int            `json:"tick_rate"`
	Rules    Rules          `json:"rules"`
	Walls    []Rect         `json:"walls"`
	Players  []ReplayPlayer `json:"players"`
	Frames   []Frame        `json:"frames"`
	Events   []Event        `json:"events"`
	Result   []PlayerResult `json:"result"`
}

func round2(v float64) float64 { return math.Round(v*100) / 100 }
func round3(v float64) float64 { return math.Round(v*1000) / 1000 }

// Frame captures the game's current state as a replay frame.
func (g *Game) Frame() Frame {
	f := Frame{T: g.Tick, Z: round2(g.Zone.R)}

	f.K = make([][]float64, len(g.Tanks))
	for i, t := range g.Tanks {
		alive := 0.0
		if t.Alive {
			alive = 1
		}
		f.K[i] = []float64{
			round2(t.X), round2(t.Y), round3(t.Hull), round3(t.Turret),
			float64(t.HP), float64(t.Reload), alive,
		}
	}

	f.S = make([][]float64, len(g.Shells))
	for i, s := range g.Shells {
		f.S[i] = []float64{float64(s.ID), float64(s.Owner), round2(s.X), round2(s.Y)}
	}

	f.B = [][]float64{}
	for _, b := range g.Bonuses {
		if b.Active {
			f.B = append(f.B, []float64{round2(b.X), round2(b.Y)})
		}
	}

	return f
}

// EncodeReplay serializes a replay as gzip-compressed JSON.
func EncodeReplay(r Replay) ([]byte, error) {
	data, err := json.Marshal(r)
	if err != nil {
		return nil, err
	}
	var buf bytes.Buffer
	zw := gzip.NewWriter(&buf)
	if _, err := zw.Write(data); err != nil {
		return nil, err
	}
	if err := zw.Close(); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

// DecodeReplay parses a gzip-compressed JSON replay.
func DecodeReplay(gz []byte) (Replay, error) {
	zr, err := gzip.NewReader(bytes.NewReader(gz))
	if err != nil {
		return Replay{}, err
	}
	defer zr.Close()
	data, err := io.ReadAll(zr)
	if err != nil {
		return Replay{}, err
	}
	var r Replay
	if err := json.Unmarshal(data, &r); err != nil {
		return Replay{}, err
	}
	return r, nil
}

File: frontend/lib/tanks/replay.ts
import { apiRaw } from '@/lib/api'

// Mirrors backend/internal/games/tanks/{rules,maps,game,replay}.go's JSON
// shapes exactly — this is the replay format written by both the server
// and `arena tanks play`, so the field names and units matter.

export interface Rules {
  width: number
  height: number
  tick_rate: number
  ticks: number
  substeps: number
  tank_radius: number
  tank_speed: number
  tank_reverse_speed: number
  hull_turn_rate: number
  turret_turn_rate: number
  reload_ticks: number
  muzzle_offset: number
  shell_speed: number
  shell_lifetime_ticks: number
  shell_damage: number
  max_hp: number
  heal_amount: number
  heal_respawn_ticks: number
  pickup_radius: number
  zone_start_tick: number
  zone_end_tick: number
  zone_start_radius: number
  zone_end_radius: number
  zone_damage_per_tick: number
}

export interface Wall { x: number; y: number; w: number; h: number }

export interface ReplayPlayer {
  slot: number
  name: string
  bot_id?: string
  version?: number
  house: boolean
  source?: string
}

// One recorded tick. k: per-slot [x, y, hull, turret, hp, reload, alive(0|1)].
// s: active shells [id, owner, x, y]. b: active bonus pickups [x, y].
export interface Frame {
  t: number
  k: number[][]
  s: number[][]
  b: number[][]
  z: number
}

// e is "shot" | "hit" | "kill" | "heal"; a is the acting tank's slot
// (-1 for "kill" means the shrinking zone did it); b is the target slot,
// d is the damage or heal amount, both present only when relevant.
export interface ReplayEvent {
  t: number
  e: string
  a: number
  b?: number
  d?: number
}

export interface PlayerResult {
  slot: number
  place: number
  kills: number
  damage: number
  death_tick: number | null
  status: string
}

export interface Replay {
  version: 1
  engine: string
  seed: number
  map: string
  tick_rate: number
  rules: Rules
  walls: Wall[]
  players: ReplayPlayer[]
  frames: Frame[]
  events: ReplayEvent[]
  result: PlayerResult[]
}

// Decompresses a gzip byte stream to the JSON replay it holds. Used by
// both fetchReplay (server response) and parseReplayFile (a local file).
async function inflateJSON(stream: ReadableStream<Uint8Array>): Promise<Replay> {
  const text = await new Response(stream.pipeThrough(new DecompressionStream('gzip'))).text()
  return JSON.parse(text) as Replay
}

export async function fetchReplay(matchId: string): Promise<Replay> {
  const res = await apiRaw(`/tanks/matches/${matchId}/replay`)
  const stream = res.body ?? (await res.blob()).stream()
  return inflateJSON(stream)
}

// Replay files are either plain JSON (as `arena tanks play` writes by
// default) or gzip-compressed JSON (the site's downloads, or a manually
// gzipped file). Sniff the gzip magic bytes rather than trusting the
// filename, since either extension can carry either encoding.
export async function parseReplayFile(file: File): Promise<Replay> {
  const buf = await file.arrayBuffer()
  const bytes = new Uint8Array(buf)
  const isGzip = bytes.length > 2 && bytes[0] === 0x1f && bytes[1] === 0x8b
  if (isGzip) return inflateJSON(new Blob([buf]).stream())
  return JSON.parse(new TextDecoder().decode(buf)) as Replay
}
